test/e2e/util/k8s/k8s_client.go

```diff
@@ -439,64 +441,107 @@ func debugPods(required int, pods []v1.Pod) {
+		// Node port first
+		hostAndPort = getNodePortAddress(c, ns, name)
+		result, err := checkExternalAddress(client, hostAndPort, ns, name, user, pass)
+		if result {
+			return result, err
 		}
-		return false, nil
+		// then ingress
+		hostAndPort := getExternalAddress(c, ns, name)
+		return checkExternalAddress(client, hostAndPort, ns, name, user, pass)
 	})
 	if err != nil {
 		panic(err.Error())
 	}
+	return hostAndPort
 }
 
-func (c ExternalK8s) CreateRoute(ns, serviceName string, portString string) int32 {
-	svc, err := c.coreClient.Services(ns).Get(serviceName, metaV1.GetOptions{})
+func checkExternalAddress(client *http.Client, hostAndPort, ns, name, user, pass string) (bool, error) {
+	req, err := http.NewRequest("GET", "http://"+hostAndPort+"/rest/default/test", nil)
+	if err != nil {
+		return false, err
+	}
+	req.SetBasicAuth(user, pass)
+	resp, err := client.Do(req)
+	if err != nil {
+		return false, nil
+	}
+	if resp.StatusCode == http.StatusNotFound || resp.StatusCode == http.StatusUnauthorized {
+		return true, nil
+	}
+	return false, nil
+}
+
+// CreateRoute exposes the port 8080 of the Infinispan cluster clusterName
+// It return the public address a string: the IP address if available else the hostname
+func (c ExternalK8s) CreateRoute(ns, clusterName string, portString string) {
+	svc, err := c.coreClient.Services(ns).Get(clusterName, metaV1.GetOptions{})
 	if err != nil {
 		panic(err.Error())
 	}
 
-	// An external route can be simply achieved with a NodePort service,
+	// An external route can be simply achieved with a LoadBalancer
 	// that has same selectors as original service.
 	route := &v1.Service{
 		TypeMeta: metaV1.TypeMeta{
 			APIVersion: "v1",
 			Kind:       "Service",
 		},
 		ObjectMeta: metaV1.ObjectMeta{
-			Name:      serviceName + "-" + portString,
+			Name:      clusterName + "-" + portString,
 			Namespace: ns,
 		},
 		Spec: v1.ServiceSpec{
-			Type:      v1.ServiceTypeNodePort,
-			Selector:  svc.Spec.Selector,
+			Type:     v1.ServiceTypeLoadBalancer,
+			Selector: svc.Spec.Selector,
 			Ports: []corev1.ServicePort{
 				{
-					Port: 8080,
+					Port:       8080,
 					TargetPort: intstr.FromInt(8080),
 				},
 			},
 		},
 	}
+	_, err = c.coreClient.Services(ns).Create(route)
+	if err != nil {
+		panic(err.Error())
+	}
+}
 
-	router, err := c.coreClient.Services(ns).Create(route)
+func getExternalAddress(c ExternalK8s, ns string, name string) string {
+	router, err := c.coreClient.Services(ns).Get(name, metaV1.GetOptions{})
 	if err != nil {
 		panic(err.Error())
 	}
+	// If the cluster exposes external IP then return it
+	if len(router.Status.LoadBalancer.Ingress) == 1 {
+		if router.Status.LoadBalancer.Ingress[0].IP != "" {
+			return router.Status.LoadBalancer.Ingress[0].IP + ":8080"
+		}
+		if router.Status.LoadBalancer.Ingress[0].Hostname != "" {
+			return router.Status.LoadBalancer.Ingress[0].Hostname + ":8080"
+		}
+	}
+	// Return empty address if nothing available
```

Comment:
Even though you return empty here, I don't see there's any checks to see if the external address is empty in the code that calls it. So, why not just throw an error if no external address can be computed?

Reply:
ok, added error there and simplified checkExternalAddress fail on error